### ISSUE TITLE
5.0.0 fails to recognize modules from within roles tasks files

### ISSUE THREAD

**Issue by luizribeiro** [ansible-community/ansible-lint#1354]
##### Summary

When a role provides a module, (for example, [`kewlfft.aur`](https://github.com/kewlfft/ansible-aur)) and that role is imported through `requirements.yml`, `ansible-lint` fails to find the module provided by the role while parsing a playbook.

This might have a similar root cause to #644. However, it only recently started reproducing after an upgrade to 5.0.0.

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
% ansible --version
ansible 2.10.5
  config file = None
  configured module search path = ['/Users/luiz/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/luiz/Library/Caches/pypoetry/virtualenvs/foo-Q681gf9M-py3.9/lib/python3.9/site-packages/ansible
  executable location = /Users/luiz/Library/Caches/pypoetry/virtualenvs/foo-Q681gf9M-py3.9/bin/ansible
  python version = 3.9.1 (default, Feb  3 2021, 07:04:48) [Clang 11.0.0 (clang-1100.0.33.17)]

% ansible-lint --version
ansible-lint 5.0.0
```

- ansible installation method: pip, through poetry
- ansible-lint installation method: pip, through poetry

##### OS / ENVIRONMENT
macOS Mojave

##### STEPS TO REPRODUCE
Setup this `requirements.yml`:

```yaml
---
roles:
  - name: kewlfft.aur
```

With a playbook as simple as:

```yaml
---
- name: Update AUR on arch boxes
  hosts: arch
  remote_user: aur_builder
  tasks:
  - name: ensure apache is at the latest version
    aur:
      upgrade: true
      aur_only: true
```

##### Desired Behaviour

It should be able to find the `aur.py` module that is provided by the `kewlfft.aur` role and `ansible-lint` should succeed.

##### Actual Behaviour

Running `ansible-lint` on this playbook yields:

```console (paste below)
% ansible-lint playbook.yml
Running ansible-galaxy install --roles-path .cache/roles -vr requirements.yml
Running ansible-galaxy collection install -p .cache/collections -vr requirements.yml
Added ANSIBLE_ROLES_PATH=.cache/roles
WARNING  Listing 1 violation(s) that are fatal
syntax-check: couldn't resolve module/action 'aur'. This often indicates a misspelling, missing collection, or incorrect module path.
playbook.yml:6:7 [WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
ERROR! couldn't resolve module/action 'aur'. This often indicates a misspelling, missing collection, or incorrect module path.

The error appears to be in '/Users/luiz/foo/playbook.yml': line 6, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  tasks:
    - name: ensure apache is at the latest version
      ^ here


You can skip specific rules or tags by adding them to your configuration file:
# .ansible-lint
warn_list:  # or 'skip_list' to silence them completely
  - syntax-check  # Ansible syntax check failed
```

**Comment by luizribeiro** [comment]
I have setup a repo that can reproduce this issue here: https://github.com/luizribeiro/ansible-lint-issue-1354

**Comment by felixfontein** [comment]
@luizribeiro if you run `ansible-playbook playbook.yml` instead of `ansible-lint playbook.yml`, does that result in the same error?

AFAIK modules from roles are only available once that role is used in Ansible. Since the playbook doesn't use the role, I don't see how Ansible should know about the module. (Also see the usage instructions: https://github.com/kewlfft/ansible-aur#ansible-galaxy - that's something you do not do in your reproducer.)

**Comment by luizribeiro** [comment]
You're right. My example is a bit too minimal.

Weirdly enough, downgrading `ansible-lint` on my main repo does help with the issue (and I can run `ansible-playbook` on it). So let me try to make the repro case actually reproduce what is going on.

**Comment by luizribeiro** [comment]
@felixfontein, I was able to get a repro that is more representative of the issue I'm seeing.

The main changes here are:

1. Moved the `aur` task into its own role, `foo`.
2. Made `roles/foo/meta/main.yml` declare a dependency on `kewlfft.aur`
3. From `playbook.yml`, made it just use the `foo` role.

This makes `ansible-lint` result in this:

```
% ansible-lint
Running ansible-galaxy install --roles-path .cache/roles -vr requirements.yml
Running ansible-galaxy collection install -p .cache/collections -vr requirements.yml
Added ANSIBLE_ROLES_PATH=roles:.cache/roles
WARNING  Listing 2 violation(s) that are fatal
meta-no-info: No 'galaxy_info' found
.:0

parser-error: couldn't resolve module/action 'aur'. This often indicates a misspelling, missing collection, or incorrect module path.
roles/foo/tasks/main.yml:2
```

(ignore the `meta-no-info`, the real problem is the `parser-error`)

Running this playbook with `ansible-playbook` works and it parses correctly by ansible.

Maybe `ansible-lint` isn't picking up the modules when the dependency is coming from my own role's `meta.yml`?

**Comment by Doudoux2** [comment]
Hello

I have the same problem with ansible-lint 5.0.0 (works ansible-lint 4.3.7)

```
parser-error: couldn't resolve module/action 'mymodule'. This often indicates a misspelling, missing collection, or incorrect module path.
src/opt/ansible/roles/myrole/tasks/main.yml:160
```

mymodule is in `src/opt/ansible/roles/myrole/library` folder

BR,

**Comment by ssbarnea** [comment]
@luizribeiro Thanks for creating this repo, it really helped me understand the bug. It is quite curious but I can confirm that linter fails where ansible has no problems with it.

In fact linter works with playbooks, but fails when it reaches the upper level role that tries to call "aur" module, which is nowhere to be found.

I am not totally surprised by this because this module is dependent on context, only pre-inclusion of the `kewlfft.aur` makes it available, something that works for the playbooks but fails for the role body.

I am not really sure how to fix it yet. If anyone has more time to look into it would be great.

**Comment by weakcamel** [comment]
+1 for the issue, also experiencing it with ansible-lint 5.0.0

An additional observation: ansible-lint seems to be ignoring settings from `ansible.cfg`.

My config is:
```
[defaults]
# ... other settings
roles_path = galaxy-roles:roles
collections_path = galaxy-roles
```

and when ansible-lint runs (and fails to find an external role), `galax-roles` is are not listed in the error message.

**Comment by ssbarnea** [comment]
@weakcamel That is mostly because current implementation does use the hardcoded list of default paths from Ansible and not the customized value. This is also caused by the lack of a reliable way to query Ansible about them, see https://github.com/ansible/ansible/issues/73644

I am not yes aware of a reliable solution for this, one workaround would be to assure that you also define environment variables using your custom paths, because the values configured there are respected.

**Comment by weakcamel** [comment]
@ssbarnea thanks for explanation. "Thumbs-upped" the referenced issue, fingers crossed it gets solved soon.

**Comment by ssbarnea** [comment]
@weakcamel lets move the discussion about config to the above ticket, it is unrelated to current bug.

I managed to get deeper into what happens but I do not have very good news.

* Ansible provides no mechanism to perform syntax check directly on a role. The only way to do it is to have a playbook as an entry point. When you do this, Ansible goes inside the role and performs syntax check on it too.
* We could disable the validation of action args on parse() and avoid reported issue when we process tasks. The downside is that this also means that people can put really invalid module there and fail to spot them.
* One workaround would be automatically mock playbooks that include detected roles in order to force ansible to do a syntax check of them, something like:

```
- hosts: localhost
   roles:
    - rolename
```

While this may not be a valid way to call any role, it should be enough to make ansible perform a minimal processing on them. It will likely test only the `main.yml` entry point, so it will not cover any `tasks/*.yml` from the role but if they are imported from within main.yml, even if this is done conditionally.
